### PR TITLE
Fixes serialization of big decimals to json (#6530)

### PR DESCRIPTION
**Pull request by mkljakubowski**
Possible fix for #6530 

Fixes serialization of BigDecimals to json in some particular case which was failing

### CODE REVIEW

File: framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala [hunk 1]
```diff
@@ -62,7 +62,7 @@ private[jackson] object JsValueSerializer extends JsonSerializer[JsValue] {
         val stripped = v.bigDecimal.stripTrailingZeros
         val raw = if (shouldWritePlain) stripped.toPlainString else stripped.toString
 
-        if (raw contains ".") json.writeTree(new DecimalNode(new JBigDec(raw)))
```
Comment (mkljakubowski) [inline framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala]: this was checking if number contains a `.` but `2e128` has no dot and still needs to be printed as BigDecimal

File: framework/src/play-json/src/test/scala/play/api/libs/json/JsonSpec.scala [hunk 2]
```diff
@@ -308,6 +308,12 @@ class JsonSpec extends org.specs2.mutable.Specification {
       jsonString must_== "0.001234"
     }
 
+    "Write BigDecimals with small negative exponents as a plain string" in {
```
Comment (gmethvin) [inline framework/src/play-json/src/test/scala/play/api/libs/json/JsonSpec.scala]: large positive exponents?

Comment (gmethvin) [inline framework/src/play-json/src/test/scala/play/api/libs/json/JsonSpec.scala]: And this is testing that the string is printed in scientific notation, not as a "plain string" (which for BigDecimals means it prints all the digits out).

**Comment by gmethvin** [comment]
Thanks. Maybe now we should just always use BigDecimal? It would make the logic a lot simpler and I'm not sure if we get much benefit from using BigInteger in some cases, as long as we know the number is being printed properly.

**Comment by mkljakubowski** [comment]
Yes, this code seems a bot overcomplicated. Let me check if using BigDecimal always will not cause any huge regression.

**Comment by mkljakubowski** [comment]
Ok, simplified code still works, but reintroduces #3784 which is still questionable. Imo makes no difference. `50 == 5e+1`

Updating jackson didn't help.

File: framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala [hunk 3]
```diff
@@ -41,30 +41,9 @@ object PlayJsonModule extends SimpleModule("PlayJson", Version.unknownVersion())
...
-        // configured with WRITE_BIGDECIMAL_AS_PLAIN, but forced as this
-        // configuration is ignored when called from ObjectMapper.valueToTree
-        val shouldWritePlain = {
-          val va = v.abs
-          va < MaxPlain && va > MinPlain
-        }
-        val stripped = v.bigDecimal.stripTrailingZeros
-        val raw = if (shouldWritePlain) stripped.toPlainString else stripped.toString
-
-        if (raw contains ".") json.writeTree(new DecimalNode(new JBigDec(raw)))
```
Comment (gmethvin) [inline framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala]: We still want to write a DecimalNode though, to preserve the formatting.

**Comment by gmethvin** [comment]
I don't think it makes a difference either but I'd rather avoid people complaining about the formatting change.

File: framework/project/Dependencies.scala [hunk 4]
```diff
@@ -25,7 +25,7 @@ object Dependencies {
     "com.fasterxml.jackson.core" % "jackson-databind",
     "com.fasterxml.jackson.datatype" % "jackson-datatype-jdk8",
     "com.fasterxml.jackson.datatype" % "jackson-datatype-jsr310"
-  ).map(_ % "2.7.6")
+  ).map(_ % "2.8.2")
```
Comment (gmethvin) [inline framework/project/Dependencies.scala]: In master this should be okay but I think minor versions of Jackson can have breaking changes, so I'd rather not upgrade in 2.5.x/2.4.x. I'd suggest we do this change in another PR so we can easily backport this one.

Comment (mkljakubowski) [inline framework/project/Dependencies.scala]: Reverted. Do you want me move this PR to 2.5.x branch? It's were the bug is after all.

Comment (schmitch) [inline framework/project/Dependencies.scala]: actually we backport things by `cherry-pick` the commit from master. and if that's not possible we manually adjust it.

File: framework/src/play-json/src/test/scala/play/api/libs/json/JsonSpec.scala [hunk 5]
```diff
@@ -342,7 +348,7 @@ class JsonSpec extends org.specs2.mutable.Specification {
 
     "Deserialize integer JsNumber as Jackson number node" in {
       val jsNum = JsNumber(new java.math.BigDecimal("50"))
-      fromJson[JsonNode](jsNum).map(_.toString) must_== JsSuccess("50")
+      fromJson[JsonNode](jsNum).map(_.toString) must_== JsSuccess("5E+1")
```
Comment (gmethvin) [inline framework/src/play-json/src/test/scala/play/api/libs/json/JsonSpec.scala]: Can you update the code so this passes again. It's fine to only use `BigDecimal

Comment (mkljakubowski) [inline framework/src/play-json/src/test/scala/play/api/libs/json/JsonSpec.scala]: Yes, I'm on it. Will be fixed today.
I also started looking in jackson for the solution but with no luck so far.

**Comment by mkljakubowski** [comment]
So the issue is I think in Jackson. This is the only piece that seems to fail:
`JacksonJson.jsValueToJsonNode(JsNumber(new java.math.BigDecimal("50"))).toString`
But if you read `JsNumber` instead of `JsonNode` everything is fine.

**Comment by gmethvin** [comment]
@mkljakubowski The reason for reading as `JsonNode` was to see how it was printed in the JSON. It seems like when Jackson parses back the number it doesn't preserve the formatting.

Can you just put back all of the old code besides the special case for integers that was causing the bug? I think that will fix it.

**Comment by mkljakubowski** [comment]
@gmethvin reverted old code
So in this case, until jackson is fixed we need to have the condition to check how the value should be printed (bigDec or bigInt).

File: framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala (hunk 1 above)
Comment (gmethvin) [inline framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala]: I'm suggesting we remove the conditional and write as:
```scala
json.writeTree(new DecimalNode(new JBigDec(raw))
```
Because of the formatting logic above, I think `raw` should be written as a plain string for a number within the range such as `50`.

Comment (mkljakubowski) [inline framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala]: Unfortunetly this breaks the tests as well. Whole issue is around jackson not beaing able to print 50 as bigdecimal and that's why we have this if to print it as bigInt.

Comment (gmethvin) [inline framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala]: Oops. I guess I didn't notice that Jackson had that special case.

File: framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala [hunk 6]
```diff
@@ -62,7 +62,7 @@ private[jackson] object JsValueSerializer extends JsonSerializer[JsValue] {
         val stripped = v.bigDecimal.stripTrailingZeros
         val raw = if (shouldWritePlain) stripped.toPlainString else stripped.toString
 
-        if (raw contains ".") json.writeTree(new DecimalNode(new JBigDec(raw)))
+        if (raw.exists(ch => ".eE".contains(ch))) json.writeTree(new DecimalNode(new JBigDec(raw)))
```
Comment (gmethvin) [inline framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala]: Given that this is common code path, maybe the most optimal way to write it is:
```scala
if (raw.indexOf('E') < 0 && raw.indexOf('.') < 0) {
  json.writeTree(new BigIntegerNode(new JBigInt(raw)))
}
json.writeTree(new DecimalNode(new JBigDec(raw)))
```

Comment (mkljakubowski) [inline framework/src/play-json/src/main/scala/play/api/libs/json/jackson/JacksonJson.scala]: fixed
